docs/extend/validation.md

```diff
@@ -31,8 +31,12 @@ If it fails validation, a `Illuminate\Validation\ValidationException` will be th
 
 Remember that you can turn an Eloquent instance into an associative array of attributes via:
 
-- The `$instance->getAttributes()` method if the instance hasn't been saved.
-- The `$instance->getDirty()` method if the instance has been saved.
+- `$instance->getAttributes()` will get all attribute values as they are currently present on the model, saved or not
+- `$instance->getDirty()` will get all attribute values that have been modified and not yet saved. It works for new or existing models
+- `$instance->getChanges()` will get all attributes that were modified and are now saved
+- `$instance->getOriginal()` will get all attribute values that were retrieved from the database
+
+Also, keep in mind that it's generally preferable to validate data before pushing it into the model instance.
```

Comment:
I think it would be nice to include a small example here. Maybe:

```php
<?php

use Illuminate\Support\Arr;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\RequestHandlerInterface;

class MyController implements RequestHandlerInterface
{
    protected $validator;
    public function __construct(UserValidator $validator) {
        $this->validator = $validator;
    }

    public function handle(ServerRequestInterface $request): ResponseInterface
    {
        $attributes = Arr::get($request->getParsedBody(), 'data.attributes', []);

        $this->validator->assertValid($attributes);

        // Do something with the attributes
        // Access them with for example Arr::get($attributes, 'email')
    }
}
```

Also worth noting somehow:

When validating an array, only keys that are present in the validator will be checked. Other will be left as-it. You should only read values from the array that were validated. To get an array of values that ran through the validator, you can use `Arr::only($attributes, array_keys($validator->getRules()))`.

A validator based on Flarum's `AbstractValidator` will only validate keys which are present in the `$data` parameter. This means `required` rules won't be checked if the data is absent. This is designed to handle API requests where each attribute will be handled separately and ignored when absent. To validate all data including missing data, you need to use a Laravel validator.